pkg/plugin/conntrack/_cprog/conntrack.c

```diff
@@ -0,0 +1,317 @@
+    } else {
+        if (CT_CONNECTION_LIFETIME_NONTCP > UINT64_MAX - now) {
+            return false;
+        }
+        WRITE_ONCE(entry->eviction_time, now + CT_CONNECTION_LIFETIME_NONTCP);
+    }
+    // We will only report this packet iff a new flag is seen for the given direction or the report interval has passed.
+    if (flags != seen_flags || now - last_report >= CT_REPORT_INTERVAL) {
+        if (direction == CT_PACKET_DIR_TX) {
+            WRITE_ONCE(entry->flags_seen_tx_dir, flags);
+            WRITE_ONCE(entry->last_report_tx_dir, now);
+        } else {
+            WRITE_ONCE(entry->flags_seen_rx_dir, flags);
+            WRITE_ONCE(entry->last_report_rx_dir, now);
+        }
+        return true;
+    }
+    return false;
+}
+
+/**
+ * Process a packet and update the connection tracking map.
+ * @arg key The key to be used to lookup the connection in the map.
+ * @arg flags The flags of the packet.
+ * @arg observation_point The point in the network stack where the packet is observed.
+ * Returns true if the packet should be report to userspace. False otherwise.
+ */
+static __always_inline __attribute__((unused)) bool ct_process_packet(struct ct_v4_key key, __u8 flags, __u8 observation_point) {    
+    // Lookup the connection in the map.
+    struct ct_entry *entry = bpf_map_lookup_elem(&retina_conntrack_map, &key);
+
+    // If the connection is found in the send direction, update the connection.
+    if (entry) {
+        return _ct_should_report_packet(entry, flags, CT_PACKET_DIR_TX, key.proto);
+    }
+    
+    // The connection is not found in the send direction. Check the reply direction by reversing the key.
+    struct ct_v4_key reverse_key;
+    __builtin_memset(&reverse_key, 0, sizeof(struct ct_v4_key));
+    _ct_reverse_key(&reverse_key, &key);
+
+    // Lookup the connection in the map based on the reverse key.
+    entry = bpf_map_lookup_elem(&retina_conntrack_map, &reverse_key);
+
+    // If the connection is found based on the reverse key, meaning that the packet is a reply packet to an existing connection.
+    if (entry) {
+        return _ct_should_report_packet(entry, flags, CT_PACKET_DIR_RX, key.proto);
+    }
+
+    // If the connection is still not found, the connection is new.
+    return _ct_handle_new_connection(key, reverse_key, flags, observation_point);
+}
+
+/**
+ * Check if a packet is a reply packet to a connection.
+ * @arg key The key to be used to check if the packet is a reply packet.
+ */
+static __always_inline __attribute__((unused)) bool ct_is_reply_packet(struct ct_v4_key key) {    
+    // Lookup the connection in the map.
+    struct ct_entry *entry = bpf_map_lookup_elem(&retina_conntrack_map, &key);
+    // We return false here because we found the connection in the send direction
+    // meaning that the packet is coming from the initiator of the connection and therefore not a reply packet.
+    return entry == NULL;
```

Comment:
nit: avoid allocating extra stack space for local variable by combining these two lines in the return statement

Reply:
```suggestion
    return bpf_map_lookup_elem(&retina_conntrack_map, &key) == NULL;
```
like this?

Reply:
yup! sorry, I'll learn the 'suggest change' feature soon :D